Deployment Guides/Provisioning-Islandora-on-Ubuntu-16.md

```diff
@@ -0,0 +1,1796 @@
+
+unzip Cantaloupe-3.4.2.zip
+
+rm -f Cantaloupe-3.4.2.zip
+
+ln -s /usr/local/Cantaloupe-3.4.2 /usr/local/cantaloupe
+
+mkdir -p /srv/cantaloupe/cache
+
+mkdir /srv/cantaloupe/log
+
+mkdir /srv/cantaloupe/home
+```
+
+#### Drush  <a id="drush"></a>
+
+Ubuntu 16.4 has Drush version 5.10, which is installed above. If that causes issues, do the following.
+
+Specify drush commit due to issues with newer drush versions and automated tests used by QA:
+```
+cd /usr/local
+
+git clone https://github.com/drush-ops/drush.git && cd drush && git checkout b9e6c8c00da0fbf1227869cdf915b0c6cea466cc
+
+ln -s /usr/local/drush/drush /usr/bin/drush
+```
+
+### Configuration <a id="configuration"></a>  
+
+#### LibreOffice <a id="libreoffice"></a>
+
+We use Systemd, which manages restarting on failure, so Monit is no longer needed.
+```
+useradd -m -d /home/libreoffice libreoffice
+
+cat > /etc/systemd/system/libreoffice.service <<END_LO
+[Unit]
+Description=Headless Libreoffice API server
+
+[Service]
+Type=simple
+User=libreoffice
+UMask=0002
+ExecStart=/usr/bin/soffice --headless --nologo --accept="socket,host=127.0.0.1,port=8100;urp"
+ExecStop=/usr/bin/killall -9 soffice.bin
+Restart=on-failure
+
+[Install]
+WantedBy=multi-user.target
+END_LO
+
+systemctl daemon-reload
+
+systemctl enable libreoffice.service
+
+systemctl start libreoffice.service
+```
+
+Verify functionality with:
+
+`unoconv --connection 'socket,host=127.0.0.1,port=8100;urp;StarOffice.ComponentContext' -f pdf <some doc in a location writeable by user libreoffice>`
+
+The following steps are required so that the `apache2` user can delete derivatives created by `libreoffice` in `/tmp`.
+
+```
+adduser www-data libreoffice
+
+mkdir /tmp/fedora
+
+chown www-data:libreoffice /tmp/fedora
+
+chmod 770 /tmp/fedora
+
+cp /usr/lib/tmpfiles.d/tmp.conf /etc/tmpfiles.d/
+
+cat >> /etc/tmpfiles.d/tmp.conf <<END_T
+# Make sure the Islandora tmp dir exists after reboot. Adjust cleaning schedule to taste.
+d /tmp/fedora 0770 www-data libreoffice -
+END_T
+```
+
+You will need to configure Drupal to account for this: set `Temporary directory` in `Configuration => Media => File system` to `/tmp/fedora`.
+
+#### Apache and PHP  <a id="apache-php"></a>
+
+Setup apache vhost:
+
+Edit `/etc/apache2/sites-available/000-default.conf`:
+
+```
+<VirtualHost *:80>
+    ServerAdmin webmaster@localhost
+
+    DocumentRoot /var/www/drupal7
+    <Directory />
+            Options FollowSymLinks
+            AllowOverride None
+    </Directory>
+    <Directory /var/www/drupal7/>
+            Options Indexes FollowSymLinks MultiViews
+            AllowOverride all
+            Order allow,deny
+            allow from all
+    </Directory>
+
+    ErrorLog ${APACHE_LOG_DIR}/error.log
+
+    # Possible values include: debug, info, notice, warn, error, crit,
+    # alert, emerg.
+    LogLevel warn
+
+    CustomLog ${APACHE_LOG_DIR}/access.log combined
+
+    # These are needed for getting Adore-Djatoka working over https
```

Comment:
I've left these in place, as in the Islandora Vagrant, and with the target of
deploying 'https' whether with Adore-Djatoka or Cantaloupe.

Reply:
Not sure I understand why is exposing these endpoints required for https?

``
/fedora/get
/fedora/services
/fedora/describe
/fedora/risearch
``
Just was curious about exposing this on a production system when they don't need to be.  I wasn't referring to the Cantaloupe or Adore-djatoka proxypass settings those would still be needed.  

For https to work a number of approaches can be taken since the usual issue is that the servlet container doesn't know how to trust the cert. If you are using a cert with a provider that is supported by Java's built in trust store you could remove "-Djavax.net.ssl.trustStore=/usr/local/fedora/server/truststore -Djavax.net.ssl.trustStorePassword=tomcat" from the Java OPTs and it might just work otherwise you could import the SSL cert chain e.g. https://tomcat.apache.org/tomcat-7.0-doc/ssl-howto.html#Importing_the_Certificate.

Reply:
The origin of these pass throughs as needed for Adore-Djatoka over https is, I believe, this
thread: https://groups.google.com/forum/?pli=1#!msg/islandora/yWlpQtL8gEU/cxF4ZmvbY8YJ

In my experience, which is limited, Cantaloupe never needed more than the standard two
line proxy pass, whether over http or https, but Adore-Djotaka just wouldn't work under https 
with those basic two.

I'll review this in the next couple of days and see if I can't make this clearer/cleaner.

Reply:
I've updated this part. I'm not sure why all those endpoints were needed, plus
I haven't actually tested them with Adore-Djatoka. But they are present in the Vagrant.
... but that's not production ready. Suggestions? Maybe it's time to drop Adore? I do
think it's important to have a working https configuration.

Reply:
I think your suggestion about the Java cert handling is probably the right way to go,
since that's probably the underlying problem. However, that doesn't explain why
https seems to work fine for Cantaloupe but not for Adore-Djatoka.

Reply:
Hi,

Vagrant is not intended to be used as a reference for a production-ready setup it is geared towards development. 

None of these should be needed in a production level setup with or without https https://github.com/Islandora-Labs/islandora_vagrant_base_box/blob/master/scripts/drupal.sh#L76-L88. You only need the one for Djatoka and Cantaloupe. 

The exposed Fedora endpoints that are exposed could be useful for someone in development who wants to bypass Islandora to hit Fedora directly (or perhaps another use case outside of Islandora that I am not aware of) otherwise they are not needed.

In Cantaloupe did you have https://github.com/discoverygarden/cantaloupe_configs/blob/master/cantaloupe.properties#L152 set to true? If so it will work with certs that are not trusted.

Reply:
Re: Cantaloupe certs checking, yes, I use that set to 'true'.

I've removed all the bogus proxy pass settings and left a note about potential issues for Adore-Djatoka.